File: terra/src/main/scala/com/dafttech/terra/game/world/tiles/TileFalling.java
package com.dafttech.terra.game.world.tiles;

import com.dafttech.terra.engine.Vector2;
import com.dafttech.terra.engine.renderer.TileRenderer;
import com.dafttech.terra.engine.renderer.TileRendererBlock;
import com.dafttech.terra.game.world.World;
import com.dafttech.terra.game.world.entities.Entity;

public abstract class TileFalling extends Tile {
    private Vector2 renderOffset = Vector2.Null();
    private float createTime = 0;

    public TileFalling() {
    }

    @Override
    public void onTick(World world, float delta) {
        super.onTick(world, delta);
        fallIfPossible(world);
        if (!renderOffset.isNull()) {
            float possSpeed = getFallSpeed(world) * delta;
            if (renderOffset.x() > 0)
                renderOffset = renderOffset.$minus(possSpeed > renderOffset.x() ? renderOffset.x() : possSpeed, 0);
            if (renderOffset.y() > 0)
                renderOffset = renderOffset.$minus(0, possSpeed > renderOffset.y() ? renderOffset.y() : possSpeed);
            if (renderOffset.x() < 0)
                renderOffset = renderOffset.$plus(possSpeed > -renderOffset.x() ? -renderOffset.x() : possSpeed, 0);
            if (renderOffset.y() < 0)
                renderOffset = renderOffset.$plus(0, possSpeed > -renderOffset.y() ? -renderOffset.y() : possSpeed);
        }
    }

    public void fall(World world, int x, int y) {
        renderOffset = renderOffset.$minus(x, y);
        world.setTile(getPosition().$plus(x, y), this, true);
    }

    public void fallIfPossible(World world) {
        if (renderOffset.isNull()) {
            if (createTime == 0) createTime = world.time();
            if (createTime + getFallDelay(world) < world.time() && world.getTile(getPosition().$plus(0, 1)).isReplacable()) {
                fall(world, 0, 1);
            }
        }
    }

    public Vector2 getRenderOffset() {
        return renderOffset;
    }

    @Override
    public TileRenderer getRenderer() {
        TileRendererBlock tileRenderer = new TileRendererBlock();
        tileRenderer.setOffset(getRenderOffset());
        return tileRenderer;
    }

    @Override
    public void onTileSet(World world) {
        fallIfPossible(world);
    }

    @Override
    public void onNeighborChange(World world, Tile changed) {
        fallIfPossible(world);
    }

    @Override
    public void onTileDestroyed(World world, Entity causer) {

    }

    @Override
    public void onTilePlaced(World world, Entity causer) {

    }

    public abstract float getFallSpeed(World world);

    public abstract float getFallDelay(World world);
}

File: terra/src/main/scala/com/dafttech/terra/game/world/tiles/TileFalling.scala
package com.dafttech.terra.game.world.tiles

import com.dafttech.terra.engine.TilePosition
import com.dafttech.terra.engine.vector.Vector2d
import com.dafttech.terra.game.world.World

abstract class TileFalling extends Tile {
  var renderOffset: Vector2d = Vector2d.Zero
  private var createTime: Float = 0

  override def onTick(delta: Float)(implicit tilePosition: TilePosition): Unit = {
    super.onTick(delta)

    fallIfPossible

    if (!renderOffset.isZero) {
      val possSpeed = getFallSpeed(tilePosition.world) * delta
      if (renderOffset.x > 0) renderOffset = renderOffset - (if (possSpeed > renderOffset.x) renderOffset.x else possSpeed, 0)
      if (renderOffset.y > 0) renderOffset = renderOffset - (0, if (possSpeed > renderOffset.y) renderOffset.y else possSpeed)
      if (renderOffset.x < 0) renderOffset = renderOffset + (if (possSpeed > -renderOffset.x) -renderOffset.x else possSpeed, 0)
      if (renderOffset.y < 0) renderOffset = renderOffset + (0, if (possSpeed > -renderOffset.y) -renderOffset.y else possSpeed)
    }
  }

  def fall(x: Int, y: Int)(implicit tilePosition: TilePosition): Unit = {
    renderOffset = renderOffset - (x, y)
    tilePosition.moveTile(tilePosition.mapPosition(_ + (x, y)))
  }

  def fallIfPossible(implicit tilePosition: TilePosition): Unit =
    if (renderOffset.isZero) {
      if (createTime == 0) createTime = tilePosition.world.time
      if (createTime + getFallDelay(tilePosition.world) < tilePosition.world.time &&
        tilePosition.world.getTile(tilePosition.pos + (0, 1)).isReplacable)
        fall(0, 1)
    }

  def getRenderOffset: Vector2d = renderOffset

  override def onTileSet(implicit tilePosition: TilePosition): Unit =
    fallIfPossible

  override def onNeighborChange(changed: TilePosition)(implicit tilePosition: TilePosition): Unit = {
    fallIfPossible
  }

  def getFallSpeed(world: World): Float

  def getFallDelay(world: World): Float
}